src/object/srv_obj.c

```diff
@@ -720,85 +1046,220 @@ ds_obj_rw_handler(crt_rpc_t *rpc)
 	struct obj_rw_out		*orwo = crt_reply_get(rpc);
 	struct ds_cont_hdl		*cont_hdl = NULL;
 	struct ds_cont			*cont = NULL;
+	struct daos_tx_handle		*dth = NULL;
+	struct daos_tx_id		*dti = NULL;
 	daos_handle_t			 ioh = DAOS_HDL_INVAL;
 	uint32_t			 map_ver = 0;
 	int				 tag;
 	bool				 update;
 	bool				 dispatch;
+	bool				 resend = false;
+	bool				 rpc_dispatched = false;
+	int				 count = 0;
 	int				 dispatch_rc = 0;
 	int				 rc;
+	int				 rc1;
 
 	D_ASSERT(orw != NULL);
 	D_ASSERT(orwo != NULL);
 	update = (opc_get(rpc->cr_opc) == DAOS_OBJ_RPC_UPDATE);
 	dispatch = update && orw->orw_shard_tgts.ca_arrays != NULL;
-	tag = dss_get_module_info()->dmi_tgt_id;
 
-	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
-		DF_U64".\n", rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
-		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
-		tag, dss_get_module_info()->dmi_xs_id, orw->orw_epoch);
+	tag = dss_get_module_info()->dmi_tgt_id;
 	rc = ds_check_container(orw->orw_co_hdl, orw->orw_co_uuid,
 				&cont_hdl, &cont);
-	if (rc)
-		goto out;
+	if (rc != 0)
+		goto reply;
 
 	if (opc_get(rpc->cr_opc) == DAOS_OBJ_RPC_UPDATE &&
 	    !(cont_hdl->sch_capas & DAOS_COO_RW)) {
 		D_ERROR("cont "DF_UUID" sch_capas "DF_U64", "
 			"NO_PERM to update.\n",
 			DP_UUID(orw->orw_co_uuid), cont_hdl->sch_capas);
-		D_GOTO(out, rc = -DER_NO_PERM);
+		D_GOTO(reply, rc = -DER_NO_PERM);
 	}
 
 	D_ASSERT(cont_hdl->sch_pool != NULL);
 	map_ver = cont_hdl->sch_pool->spc_map_version;
+
+	D_DEBUG(DB_TRACE, "rpc %p opc %d "DF_UOID" dkey %d %s tag/xs %d/%d eph "
+		DF_U64", pool ver %u/%u with "DF_DTI".\n",
+		rpc, opc_get(rpc->cr_opc), DP_UOID(orw->orw_oid),
+		(int)orw->orw_dkey.iov_len, (char *)orw->orw_dkey.iov_buf,
+		tag, dss_get_module_info()->dmi_xs_id, orw->orw_epoch,
+		orw->orw_map_ver, map_ver, DP_DTI(&orw->orw_dti));
+
 	if (orw->orw_map_ver < map_ver) {
 		D_DEBUG(DB_IO, "stale version req %d map_version %d\n",
 			orw->orw_map_ver, map_ver);
 		if (update && dispatch)
-			D_GOTO(out, rc = -DER_STALE);
+			D_GOTO(reply, rc = -DER_STALE);
+	} else if (orw->orw_map_ver > map_ver) {
+		/* XXX: client has newer pool map than the server.
+		 *	Refresh the pool map on the server?
+		 */
+	}
+
+	if (!update ||
+	    daos_oc_echo_type(daos_obj_id2class(orw->orw_oid.id_pub)))
+		goto local;
+
+	if (orw->orw_flags & ORF_FROM_LEADER) {
+		if ((orw->orw_flags & ORF_DTX_DISABLED && srv_enable_dtx) ||
+		    (!(orw->orw_flags & ORF_DTX_DISABLED) && !srv_enable_dtx)) {
```

Comment:
(orw->orw_flags & ORF_DTX_DISABLED == 0) != srv_enable_dtx ?

Reply:
"orw->orw_flags & ORF_DTX_DISABLED == 0" means remote replica (leader) want to enable DTX. "srv_enable_dtx" mean local server want to enable DTX. If local replica and remote replica have different configuration for the DTX, then return failure.

Reply:
right, just seems the 2 lines check can be replaced by 1 line, not a big deal.

Reply:
1108 line means leader disabled DTX but local replica enabled DTX.
1109 line means leader enabled DTX but local replica disabled DTX.

Reply:
minor, both lines covered by "(orw->orw_flags & ORF_DTX_DISABLED == 0) != srv_enable_dtx "

Reply:
OK